### PR TITLE
add API `const_new_unchecked` to `Identifier`

### PR DESCRIPTION
**Pull request by lightsing**
## What's this PR do
adds the new function `const_new_unchecked` to the Identifier struct, which enables the creation of an identifier with a constant string without any validation check.

## Motivation
Consider following codes:
```rust
mod identifiers {
    const PERSON: Identifier = unsafe { Identifier::const_new_unchecked("person") };
    const MOVIE: Identifier = unsafe { Identifier::const_new_unchecked("movie") };
}
```

This allows to create constant identifiers for encapsulation and abstraction.

## Changes
1. The addition of a new `MayStatic` enum type, which has two possible variants:
    * `Static`: A variant that holds a string reference to a constant `&'static str`.
    * `Interned`: A variant that holds an `Intern<String>` instance.
2. The modification of the existing `Identifier` struct to use the `MayStatic` enum instead of an `Intern<String>` instance to hold the value.

### CODE REVIEW

**Comment by ysimonson** [comment]
Support for const would be nice, but I'm not sure it's worth the cost of branching on all `Identifier` methods, since they're used so heavily. Could you outline the use case? Maybe there's an alternative.

As an aside, it is sort of silly that `Intern::new` isn't const, given it makes the internal value are `'static` anyways. I think a rigorous fix to support const would need to be applied there. But maybe there's a reason why it's not possible.

**Comment by lightsing** [comment]
> Support for const would be nice, but I'm not sure it's worth the cost of branching on all `Identifier` methods, since they're used so heavily. Could you outline the use case? Maybe there's an alternative.
> 
> As an aside, it is sort of silly that `Intern::new` isn't const, given it makes the internal value `'static` anyways. I think a rigorous fix to support const would need to be applied there. But maybe there's a reason why it's not possible.

You are right about the "branching" thing. Maybe we can find a better way.

The `Intern::new` can't be const since it need to allocate memory in the internal `HashSet`.

I'm new to graph db, so why the `Identifier` is usually build in runtime? As I'm trying out indradb, I pre-defined all possible `Identifier` in a mod via `once_cell::sync::Lazy` and use only their references.
**IF** most common way to use `Identifier` is pre-define, may be we can use a better way to hold the internal string.

**Comment by lightsing** [comment]
The crate internment is not designed for build from 'static, since it wants to manage all thing it holds and remove duplicate.

So, I remove the internment dep and impl a special version only holds 'static str for indradb.

**Comment by lightsing** [comment]
p.s. I suggest to change `lazy_static` to `once_cell`

https://github.com/indradb/indradb/blob/a774da9585a02eba07cd026ac130dda7bd565d89/lib/src/util.rs#L19-L21

**Comment by ysimonson** [comment]
> I'm new to graph db, so why the Identifier is usually build in runtime? As I'm trying out indradb, I pre-defined all possible Identifier in a mod via once_cell::sync::Lazy and use only their references.

I suppose runtime would be more common, since it's necessary when called via gRPC. That said, supporting both would be nice.

> p.s. I suggest to change lazy_static to once_cell

Makes sense, happy to take a PR with this change.

File: lib/src/models/identifiers.rs [hunk 1]
```diff
@@ -1,17 +1,58 @@
...
 /// letters, numbers, dashes and underscores. This is used for vertex and edge
 /// types, as well as property names.
 #[derive(Eq, PartialEq, Copy, Clone, Debug, Hash, Ord, PartialOrd)]
-pub struct Identifier(pub(crate) Intern<String>);
+pub struct Identifier(pub(crate) Intern);
+
+static INTERNAL_CONTAINERS: Lazy<Mutex<HashSet<&'static str>>> = Lazy::new(|| Mutex::new(HashSet::new()));
+
+#[derive(Eq, PartialEq, Copy, Clone, Debug, Hash, Ord, PartialOrd)]
+pub(crate) struct Intern {
```
Comment (ysimonson) [inline lib/src/models/identifiers.rs:21]: Needs a less vague name, maybe `InternedString`?

File: lib/src/models/identifiers.rs [hunk 2]
```diff
@@ -1,17 +1,58 @@
...
 use serde::{Deserialize, Deserializer, Serialize, Serializer};
 
 /// A string that must be less than 256 characters long, and can only contain
 /// letters, numbers, dashes and underscores. This is used for vertex and edge
 /// types, as well as property names.
 #[derive(Eq, PartialEq, Copy, Clone, Debug, Hash, Ord, PartialOrd)]
-pub struct Identifier(pub(crate) Intern<String>);
+pub struct Identifier(pub(crate) Intern);
+
+static INTERNAL_CONTAINERS: Lazy<Mutex<HashSet<&'static str>>> = Lazy::new(|| Mutex::new(HashSet::new()));
```
Comment (ysimonson) [inline lib/src/models/identifiers.rs:18]: Could you rename to `INTERNED_STRINGS`?

**Comment by ysimonson** [comment]
This seems to be a fair bit slower than internment. The microbenchmarks don't test for `Identifier` overhead, but nevertheless seeing an impact:

```
MASTER
test memory::bench_bulk_insert    ... bench:   8,422,419 ns/iter (+/- 1,077,473)
test memory::bench_create_edge    ... bench:          51 ns/iter (+/- 3)
test memory::bench_create_vertex  ... bench:         199 ns/iter (+/- 13)
test memory::bench_get_edge_count ... bench:          91 ns/iter (+/- 5)
test memory::bench_get_edges      ... bench:         227 ns/iter (+/- 11)
test memory::bench_get_vertices   ... bench:         209 ns/iter (+/- 10)

PR #282
test memory::bench_bulk_insert    ... bench:   9,451,130 ns/iter (+/- 1,035,730)
test memory::bench_create_edge    ... bench:          59 ns/iter (+/- 6)
test memory::bench_create_vertex  ... bench:         209 ns/iter (+/- 22)
test memory::bench_get_edge_count ... bench:         111 ns/iter (+/- 83)
test memory::bench_get_edges      ... bench:         254 ns/iter (+/- 52)
test memory::bench_get_vertices   ... bench:         235 ns/iter (+/- 37)
```

Not sure, but I wonder if the overhead is coming from doing full string comparisons rather than cheap pointer comparisons.

**Comment by lightsing** [comment]
I got a different result on wsl Ubuntu 22.04.2 LTS.

```
MASTER
test memory::bench_bulk_insert    ... bench:  17,438,960 ns/iter (+/- 18,438,881)
test memory::bench_create_edge    ... bench:          75 ns/iter (+/- 121)
test memory::bench_create_vertex  ... bench:         653 ns/iter (+/- 418)
test memory::bench_get_edge_count ... bench:          90 ns/iter (+/- 59)
test memory::bench_get_edges      ... bench:         228 ns/iter (+/- 223)
test memory::bench_get_vertices   ... bench:         226 ns/iter (+/- 242)

PR #282
test memory::bench_bulk_insert    ... bench:   6,139,715 ns/iter (+/- 779,448)
test memory::bench_create_edge    ... bench:          36 ns/iter (+/- 11)
test memory::bench_create_vertex  ... bench:         204 ns/iter (+/- 34)
test memory::bench_get_edge_count ... bench:          54 ns/iter (+/- 8)
test memory::bench_get_edges      ... bench:         119 ns/iter (+/- 20)
test memory::bench_get_vertices   ... bench:         109 ns/iter (+/- 11)
```

**Comment by lightsing** [comment]
```
performance change after multi-slot
test memory::bench_bulk_insert    ... bench:   5,761,460 ns/iter (+/- 522,578)
test memory::bench_create_edge    ... bench:          36 ns/iter (+/- 5)
test memory::bench_create_vertex  ... bench:         199 ns/iter (+/- 44)
test memory::bench_get_edge_count ... bench:          50 ns/iter (+/- 9)
test memory::bench_get_edges      ... bench:         129 ns/iter (+/- 17)
test memory::bench_get_vertices   ... bench:         107 ns/iter (+/- 15)
```

**Comment by ysimonson** [comment]
Interesting. I think this warrants some more direct benchmarking. Will do so in #284.

**Comment by lightsing** [comment]
There's no noticable performance change.
```
MASTER
test benches::benches::bench_ident_comparison      ... bench:           9 ns/iter (+/- 2)
test benches::benches::bench_ident_new             ... bench:       2,407 ns/iter (+/- 471)
test benches::benches::bench_ident_new_unchecked   ... bench:       1,905 ns/iter (+/- 328)
test benches::benches::bench_ident_renew           ... bench:         500 ns/iter (+/- 95)
test benches::benches::bench_ident_renew_unchecked ... bench:          63 ns/iter (+/- 13)

PR #282 mutli-slot disabled
test benches::benches::bench_ident_comparison      ... bench:          14 ns/iter (+/- 4)
test benches::benches::bench_ident_new             ... bench:       2,312 ns/iter (+/- 504)
test benches::benches::bench_ident_new_unchecked   ... bench:       1,746 ns/iter (+/- 456)
test benches::benches::bench_ident_renew           ... bench:         515 ns/iter (+/- 93)
test benches::benches::bench_ident_renew_unchecked ... bench:         104 ns/iter (+/- 14)
```
`multi-slot` is optimized for parallel `new`, but current benchmark code cannot bench this.

**Comment by ysimonson** [comment]
I'm still seeing a fairly noticeable performance degradation on a macbook pro / intel i5:

```
MASTER
test benches::benches::bench_ident_comparison      ... bench:          11 ns/iter (+/- 1)
test benches::benches::bench_ident_new             ... bench:       1,955 ns/iter (+/- 311)
test benches::benches::bench_ident_new_unchecked   ... bench:       1,427 ns/iter (+/- 193)
test benches::benches::bench_ident_renew           ... bench:         346 ns/iter (+/- 103)
test benches::benches::bench_ident_renew_unchecked ... bench:          92 ns/iter (+/- 15)

PR #282
test benches::benches::bench_ident_comparison      ... bench:          23 ns/iter (+/- 7)
test benches::benches::bench_ident_new             ... bench:       2,323 ns/iter (+/- 804)
test benches::benches::bench_ident_new_unchecked   ... bench:       1,665 ns/iter (+/- 994)
test benches::benches::bench_ident_renew           ... bench:         388 ns/iter (+/- 41)
test benches::benches::bench_ident_renew_unchecked ... bench:         156 ns/iter (+/- 74)
test memory::bench_bulk_insert                     ... bench:  10,255,757 ns/iter (+/- 4,132,138)
test memory::bench_create_edge                     ... bench:          59 ns/iter (+/- 10)
```

**Comment by lightsing** [comment]
The difference in measurements with similar errors is not statistically significant, and the measurement results have no statistical significance.

**Comment by ysimonson** [comment]
This is interesting work, but I don't think it's worthwhile to merge.

`bench_ident_comparison` implies to me that there is some sort of performance regression. I suspect this has to do with equality checks -- internment likely uses cheap pointer checks, versus this derives `Eq` which will do full string equality. More broadly, though, this introduces non-trivial complexity to IndraDB, which has downsides for maintainability. More tests would be necessary (e.g. to ensure constructing two identifiers with the same value reuses memory). And the multi-slot feature adds complexity for end-users (how should they decide whether to use it or not?)

Overall, this seems like functionality that's well suited to move behind a library like internment. While we don't get const new with that library, I feel that the utility of that is limited, since nothing else in IndraDB supports const new anyways.

**Pull request closed without merging**